### PR TITLE
add redemption fee and fee tests

### PR DESCRIPTION
**Pull request by dpaiton**

### CODE REVIEW

**Review by wakamex — APPROVED** [review]
lgtm, you sexy motherfucker

**Comment by wakamex** [comment]
what solidity does:
```
        HyperdriveBase.sol:
            _calculateFeesOutGivenSharesIn: (input is shares, so we're calculating the fee on the input)
            (used in HyperdriveLong.sol)
                curve fee = ((1 / p) - 1) * d_z * c * t * phi_curve
                    ((1 / p) - 1) = (1 - p) / p
                curve fee = (1 - p) * d_z * c * t * phi_curve
                    units are:
                    d_z is in shares
                    multiplying by c makes it base
                     ?? dividing by p makes it $1
                     ?? multiplying by $ (1 - p) makes it in bonds??
                govCurveFee = d_z * (curve_fee/d_y) * c * gov_fee_percent
                    d_z/d_y is the price, so this simplifies to:
                govCurveFee = p * curve_fee * c * gov_fee_percent
                flat fee = c * d_z * (1 - t) * phi_flat

        HyperdriveBase.sol:
            _calculateFeesInGivenBondsOut: (input is shares, so we're calculating the fee on the input)
            (used in HyperdriveShort.sol)
                curve fee = ((1 - p) * d_y * t * phi_curve) / c
                    units are:
                    d_y is in bonds
                    multiplying it by (1-p) makes it base
                    dividing by c makes it shares
                gov_curve_fee = curve_fee * gov_fee_percent
                flat fee = (d_y * (1 - t) * phi_flat) / c

            _calculateFeesOutGivenBondsIn: (output is shares, so we're calculating the fee on the output)
            (used in HyperdriveLong.sol and HyperdriveShort.sol)
                curve fee = ((1 - p) * d_y * t * phi_curve) / c
                gov_curve_fee = curve_fee * gov_fee_percent
                flat fee = (d_y * (1 - t) * phi_flat) / c


        HyperriveLong.sol calls:
            - _calculateFeesOutGivenSharesIn DIFF
            - _calculateFeesOutGivenBondsIn  same
        HyperdriveShort.sol calls:
            - _calculateFeesOutGivenBondsIn  same
            - _calculateFeesInGivenBondsOut  DIFF
```

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 1]
```diff
@@ -170,9 +172,7 @@ def calc_open_short(
         address=wallet_address,
         balance=-types.Quantity(amount=max_loss, unit=types.TokenType.BASE),
         shorts={
-            market.latest_checkpoint_time: wallet.Short(
-                balance=bond_amount, open_share_price=market.market_state.share_price
-            )
+            market.block_time.time: wallet.Short(balance=bond_amount, open_share_price=market.market_state.share_price)
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:175]: I don't think this is right -- I think it should be latest_checkpoint_time

File: elfpy/markets/hyperdrive/hyperdrive_market.py [hunk 2]
```diff
@@ -107,8 +108,11 @@ def __setitem__(self, key, value):
     # fee percents
     trade_fee_percent: float = field(default=0.0)
     redemption_fee_percent: float = field(default=0.0)
+    governance_fee_percent: float = field(default=0.0)
 
-    # The amount of longs that are still open.
+    # governance fees that haven't been collected yet denominated in shares
+    gov_fees_accrued: Decimal = field(default=Decimal(0))
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_market.py:114]: Do the tests fail if this is a float? I'd like to keep it consistent with the other parameters and we will be getting rid of Decimal in a the upcoming precision PR.

File: elfpy/markets/hyperdrive/hyperdrive_market.py [hunk 3]
```diff
@@ -256,11 +260,6 @@ def __init__(
         self.position_duration.freeze()  # pylint: disable=no-member # type: ignore
         super().__init__(pricing_model=pricing_model, market_state=market_state, block_time=block_time)
 
-    @property
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_market.py:259]: Why did you delete this?

File: elfpy/pricing_models/hyperdrive.py [hunk 4]
```diff
@@ -257,39 +262,48 @@ def calc_out_given_in(
...
-        normalized_time = Decimal(time_remaining.normalized_time)
+        curve_portion = Decimal(time_remaining.normalized_time)
+        flat_portion = 1 - curve_portion
         share_price = Decimal(market_state.share_price)
-        d_bonds = in_amount * (1 - normalized_time)
+        d_bonds = in_amount * flat_portion  # whether in_.unit is base or pt, at maturity d_bonds = d_base
         d_shares = d_bonds / share_price
         # Redeem the matured bonds 1:1 and simulate these updates hitting the reserves.
-        market_state = market_state.copy()  # don't want to modify the actual market state
+        market_state_simulated = market_state.copy()  # don't want to modify the actual market state
```
Comment (dpaiton) [inline elfpy/pricing_models/hyperdrive.py:276]: `market_state_simulated = market_state.copy()` causes us to store `market_state` twice in memory, while the other version ` market_state = market_state.copy() ` does not. If we don't _need_ two versions, then let's leave it how it was.

File: elfpy/pricing_models/hyperdrive.py [hunk 5]
```diff
@@ -257,39 +262,48 @@ def calc_out_given_in(
...
         elif in_.unit == types.TokenType.PT:
-            market_state.share_reserves -= float(d_shares)
-            market_state.bond_reserves += float(d_bonds)
+            market_state_simulated.share_reserves -= float(d_shares)
+            market_state_simulated.bond_reserves += float(d_bonds)
         else:
             raise AssertionError(
                 "pricing_models.calc_out_given_in: ERROR: "
-                f"Expected in_.unit to be {types.TokenType.BASE} or {types.TokenType.PT}, not {in_.unit}!"
+                f"Expected in_.unit to be {types.TokenType.PT}, not {in_.unit}!"
```
Comment (dpaiton) [inline elfpy/pricing_models/hyperdrive.py:286]: What was wrong with the old message? BASE or PT are allowed.

File: elfpy/pricing_models/yieldspace.py [hunk 6]
```diff
@@ -44,7 +45,7 @@ def calc_lp_out_given_tokens_in(
         rate: float,
         market_state: hyperdrive_market.MarketState,
         time_remaining: time.StretchedTime,
-    ) -> tuple[float, float, float]:
+    ) -> Tuple[float, float, float]:
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:48]: Why did you change this? I think the lower-case `tuple` is correct, and the upper-case `Tuple` is being deprecated.

File: elfpy/pricing_models/yieldspace.py [hunk 7]
```diff
@@ -650,10 +632,51 @@ def calc_out_given_in(
...
+        """
+        phi_curve = Decimal(market_state.trade_fee_percent)
+        governance_fee_percent = Decimal(market_state.governance_fee_percent)
+        share_price = Decimal(market_state.share_price)
+        curve_fee = Decimal(np.nan)
+
+        # multiplying by time remaining is not required below, since the amount
+        # passed into this function is already scaled down to the unmatured portion
+        if func == "_calculateFeesOutGivenSharesIn":
+            d_z = amount
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:660]: change to d_shares

File: elfpy/pricing_models/yieldspace.py [hunk 8]
```diff
@@ -650,10 +632,51 @@ def calc_out_given_in(
...
+            # curve fee = ((1 / p) - 1) * d_z * c * t * phi_curve
+            curve_fee = abs(1 - spot_price) * phi_curve * d_z
+        elif func == "_calculateFeesInGivenBondsOut":
+            d_y = amount  # in bonds
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:664]: change to d_bonds

File: elfpy/pricing_models/yieldspace.py [hunk 9]
```diff
@@ -650,10 +632,51 @@ def calc_out_given_in(
...
+            # curve fee = ((1 - p) * d_y * t * phi_curve) / c
+            curve_fee = ((1 - spot_price) * d_y * phi_curve) / share_price
+        elif func == "_calculateFeesOutGivenBondsIn":
+            d_y = amount  # in bonds
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:668]: change to d_bonds

**Review by wakamex — APPROVED** [review]

File: elfpy/pricing_models/hyperdrive.py [hunk 10]
```diff
@@ -132,18 +144,22 @@ def calc_in_given_out(
...
+            hyperdrive=market_state,  # pass in the real market state to apply the gov fee
         )
 
         # Compute flat part with fee
         flat_without_fee = out_amount * (1 - normalized_time)
-        redemption_fee = flat_without_fee * Decimal(market_state.redemption_fee_percent)
-        flat_with_fee = flat_without_fee + redemption_fee
+        total_redemption_fee, _, _ = self.calc_flat_fee_split_and_apply_gov_fee(
+            amount=flat_without_fee,
+            market_state=market_state,  # pass in the real market state to apply the gov fee
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/pricing_models/hyperdrive.py]: actually i removed this function completely, since it only executes 3 lines of code

File: tests/solidity/test_fees.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+
+import elfpy.agents.agent as agent
+import elfpy.markets.hyperdrive.hyperdrive_market as hyperdrive_market
+import elfpy.pricing_models.hyperdrive as hyperdrive_pm
+import elfpy.time as time
+import elfpy.types as types
+
+# pylint: disable=duplicate-code
+
+ONEE18 = Decimal(10) ** -18
```
[the remark this answers is not in the document]
Comment (wakamex) [reply tests/solidity/test_fees.py]: actually removed since ether function is no more

File: elfpy/pricing_models/yieldspace.py (hunk 6 above)
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py:48]: oops I thought I saw way more higher case Tuples in our code, i'll change it back

File: elfpy/pricing_models/hyperdrive.py (hunk 5 above)
Comment (wakamex) [reply elfpy/pricing_models/hyperdrive.py:286]: yeah i forgot to undo this change, changing back

File: elfpy/markets/hyperdrive/hyperdrive_market.py (hunk 3 above)
Comment (wakamex) [reply elfpy/markets/hyperdrive/hyperdrive_market.py:259]: I saw it wasn't used anywhere and it's just returning a variable without any calculation

File: elfpy/markets/hyperdrive/hyperdrive_market.py (hunk 2 above)
Comment (wakamex) [reply elfpy/markets/hyperdrive/hyperdrive_market.py]: nope, changed it to float

File: examples/notebooks/vault_tracker.ipynb [hunk 12]
```diff
@@ -158,8 +158,8 @@
         "config.num_trading_days = 365 # Number of simulated trading days, default is 180\n",
         "config.num_position_days = config.num_trading_days # term length\n",
         "config.num_blocks_per_day = 1 #7200 # Blocks in a given day (7200 means ~12 sec per block)\n",
-        "config.trade_fee_percent = 0#0.10 # fee percent collected on trades\n",
-        "config.redemption_fee_percent = 0.000 # 5 bps\n",
+        "config.trade_fee_mult = 0#0.10 # fee percent collected on trades\n",
+        "config.redemption_fee_mult = 0.000 # 5 bps\n",
```
[the remark this answers is not in the document]
Comment (wakamex) [reply examples/notebooks/vault_tracker.ipynb]: default is 0 so it's safe

File: elfpy/markets/hyperdrive/hyperdrive_market.py (hunk 3 above)
Comment (dpaiton) [reply elfpy/markets/hyperdrive/hyperdrive_market.py:259]: It is being used in the bots_on_solidity notebook -- please leave it in.

File: elfpy/pricing_models/hyperdrive.py [hunk 13]
```diff
@@ -107,17 +106,18 @@ def calc_in_given_out(
             The result of performing the trade
         """
 
+        # pylint: disable=too-many-locals
+
         # Calculate some common values up front
         out_amount = Decimal(out.amount)
-        normalized_time = Decimal(time_remaining.normalized_time)
+        curve_portion = Decimal(time_remaining.normalized_time)
```
Comment (dpaiton) [inline elfpy/pricing_models/hyperdrive.py:113]: I find these variable names to be more confusing because it no indicates that it is an indicator of time until maturity. What about switching `normalized_time` to `time_remaining` and leaving `(1-time_remaining)`?

File: tests/solidity/test_fees.py [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+
+
+def get_gov_fees_accrued(test, market_state=None) -> float:
+    """Get the amount of gov fees that have accrued in the market state"""
+    if market_state:
+        return market_state.gov_fees_accrued * market_state.share_price
+    return test.hyperdrive.market_state.gov_fees_accrued * test.hyperdrive.market_state.share_price
+
+
+def warp(test, time_delta):
```
Comment (dpaiton) [inline tests/solidity/test_fees.py:99]: Get rid of this -- use `block_time.tick()`

File: tests/solidity/test_fees.py [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+    """Move time forward by time_delta and update the share price to simulate interest"""
+    test.hyperdrive.block_time.set_time(test.hyperdrive.block_time.time + time_delta)
+    test.hyperdrive.market_state.share_price = test.market_state_before_open.share_price * (
```
Comment (dpaiton) [inline tests/solidity/test_fees.py:102]: verify that this is indeed exactly the same every time it is done in Solidity

File: elfpy/pricing_models/yieldspace.py [hunk 16]
```diff
@@ -650,10 +632,51 @@ def calc_out_given_in(
...
+        governance_fee_percent = Decimal(market_state.governance_fee_percent)
+        share_price = Decimal(market_state.share_price)
+        curve_fee = Decimal(np.nan)
+
+        # multiplying by time remaining is not required below, since the amount
+        # passed into this function is already scaled down to the unmatured portion
+        if func == "_calculateFeesOutGivenSharesIn":
+            d_shares = amount
+            # curve fee = ((1 / p) - 1) * d_z * c * t * phi_curve
+            curve_fee = abs(1 - spot_price) * phi_curve * d_shares
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/pricing_models/yieldspace.py:662]: Agreed lets move these back into `calc_*_given_*`

File: elfpy/pricing_models/yieldspace.py [hunk 17]
```diff
@@ -396,23 +385,16 @@ def calc_in_given_out(
...
-            #
-            # fee = (1 - p) * phi * d_y
-            fee = (1 - spot_price) * trade_fee_percent * d_bonds
-            logging.debug(
-                ("fee = (1 - spot_price) * _fee_percent * d_bonds = (1 - %g) * %g * %g = %g"),
-                spot_price,
-                trade_fee_percent,
-                d_bonds,
-                fee,
+            curve_fee, gov_curve_fee = self.calc_curve_fee_split(
```
Comment (sentilesdal) [inline elfpy/pricing_models/yieldspace.py:388]: let's keep the fees calcs here inline and just add the gov fee part

File: tests/solidity/test_fees.py [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+    test.assertEqual(gov_fees_after_collection, 0)
+
+    gov_balance_after = test.gary.wallet.balance.amount
+    # ensure that Governance Gary's balance has increased
+    test.assertGreater(gov_balance_after, gov_balance_before_open_short)
+    # ensure that Governance Gary got the exaxt fees expected
+    test.assertAlmostEqual(gov_balance_after, gov_fees_after_close_short, delta=1e-16 * test.trade_amount)
+
+
+def get_all_the_fees(test: TestFees, func: str) -> Tuple[float, float, float, float]:
```
Comment (dpaiton) [inline tests/solidity/test_fees.py:223]: move this to the top

File: tests/solidity/test_fees.py [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+        time_stretch=test.hyperdrive.position_duration.time_stretch,
+        normalizing_constant=test.hyperdrive.position_duration.normalizing_constant,
+    ).normalized_time
+
+    # calculate curve fee
+    spot_price_adjusted = test.hyperdrive.spot_price
+    if func == "_calculateFeesOutGivenSharesIn":  # input is in shares
+        spot_price_adjusted = 1 / test.hyperdrive.spot_price  # when given shares, invert price
+
+    curve_fee, gov_curve_fee = test.hyperdrive.pricing_model.calc_curve_fee_split(
```
Comment (dpaiton) [inline tests/solidity/test_fees.py:242]: should be able to get gov fee from trade breakdown in `calc_*_given_*`

File: elfpy/markets/hyperdrive/hyperdrive_market.py (hunk 3 above)
Comment (wakamex) [reply elfpy/markets/hyperdrive/hyperdrive_market.py]: put it back in 7419cf280eaa85048f4c55732913aa1a4716630a

File: elfpy/pricing_models/hyperdrive.py (hunk 13 above)
Comment (wakamex) [reply elfpy/pricing_models/hyperdrive.py]: done in f10f1e770d8559883fdf267802c77f0b6a9ef96e

File: elfpy/simulators/simulators.py [hunk 20]
```diff
@@ -121,10 +121,12 @@ class Config:
...
     # Must be "Hyperdrive", or "YieldSpace"
     pricing_model_name: str = field(default="Hyperdrive")
-    # LP fee factor (decimal) to charge for trades
+    # LP fee factor (float) to charge for trades
     trade_fee_percent: float = field(default=0.05)
-    # LP fee factor (decimal) to charge for redemption
+    # LP fee factor (float) to charge for redemption
     redemption_fee_percent: float = field(default=0.05)
+    # minimum fee for governance
+    governance_fee_percent: float = field(default=0)
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/simulators/simulators.py:129]: to separate the PRs so they only do 1 thing at a time 😂

[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/simulators/simulators.py:129]: ( `percent` is consistent with the naming convention that is live in the codebase as of this PR... no leaking information about future PRs here! )

File: elfpy/pricing_models/yieldspace.py (hunk 17 above)
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py]: done in ebe871de8cd655cc7403dfa24ee00e25b9eb2fe7

File: elfpy/pricing_models/yieldspace.py [hunk 21]
```diff
@@ -332,27 +332,16 @@ def calc_in_given_out(
...
-                spot_price,
-                trade_fee_percent,
-                share_price,
-                d_shares,
-                fee,
+            curve_fee, gov_curve_fee = self.calc_curve_fee_split(
+                amount=d_shares * share_price,  # taking a cut of base
+                spot_price=1 / spot_price,  # when given shares, invert price
+                market_state=market_state,
+                func="_calculateFeesOutGivenSharesIn",
```
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py]: removed in ebe871de8cd655cc7403dfa24ee00e25b9eb2fe7

File: elfpy/pricing_models/yieldspace.py (hunk 17 above)
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py]: removed in ebe871de8cd655cc7403dfa24ee00e25b9eb2fe7

File: elfpy/pricing_models/yieldspace.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py:651]: inlined in ebe871de8cd655cc7403dfa24ee00e25b9eb2fe7
keeping this function to match with tests, but will look at removing it next

File: tests/solidity/test_fees.py (hunk 15 above)
Comment (wakamex) [reply tests/solidity/test_fees.py:102]: all 3 instances of setSharePrice in FeeTest.t.sol are identical:
```
hyperdrive.setSharePrice(
            getPoolInfo().sharePrice.mulDown(
                FixedPointMath.ONE_18 + apr.mulDown(timeDelta)
            )
        );
```

File: tests/solidity/test_fees.py (hunk 19 above)
Comment (wakamex) [reply tests/solidity/test_fees.py]: done in 00338cdf842d002780943bd00928077a6622826e

File: elfpy/pricing_models/yieldspace.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py]: removed in 00338cdf842d002780943bd00928077a6622826e

File: elfpy/pricing_models/yieldspace.py [hunk 22]
```diff
@@ -332,27 +333,15 @@ def calc_in_given_out(
...
-                    "fee = ((1 / spot_price) - 1) * _fee_percent * share_price * d_shares = "
-                    "((1 / %g) - 1) * %g * %g * %g = %g"
-                ),
-                spot_price,
-                trade_fee_percent,
-                share_price,
-                d_shares,
-                fee,
-            )
+            print(f"{without_fee_or_slippage=}")
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:336]: no prints

File: elfpy/pricing_models/trades.py [hunk 23]
```diff
@@ -33,7 +33,20 @@ class TradeBreakdown:
     without_fee_or_slippage: float
     with_fee: float
     without_fee: float
-    fee: float
+    curve_fee: float
+    gov_curve_fee: float
+    redemption_fee: float = 0.0
+    gov_redemption_fee: float = 0.0
```
Comment (dpaiton) [inline elfpy/pricing_models/trades.py:39]: Where is `gov_redemption_fee` used?

File: elfpy/markets/hyperdrive/hyperdrive_market.py [hunk 24]
```diff
@@ -289,6 +292,25 @@ def spot_price(self) -> float:
             time_remaining=self.position_duration,
         )
 
+    def collect_gov_fee(self, governance_wallet: wallet.Wallet):
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_market.py:295]: From what I can tell, this function is only used in the tests, so it should live there. This keeps code closer to where it is used, and also in general the market should not be directly updating an agent's wallet. Alternatively, in the test you should be able to get the fee information via the agent_delta Wallet and then update it at the test level. Then you can delete `collect_gov_fee` all together. That's my preference.

File: elfpy/pricing_models/yieldspace.py [hunk 25]
```diff
@@ -566,16 +546,13 @@ def calc_out_given_in(
                 share_price=share_price,
                 init_share_price=init_share_price,
             )
-            # The fees are calculated as the difference between the bonds
-            # received without slippage and the base paid times the fee
-            # percentage. This can also be expressed as:
-            #
-            # ((1 / p) - 1) * phi * c * dz
-            fee = ((1 / spot_price) - 1) * trade_fee_percent * share_price * d_shares
+            curve_fee = abs(in_amount - without_fee_or_slippage) * trade_fee_percent
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:549]: shouldn't this be `curve_fee = abs(without_fee_or_slippage - in_amount) * trade_fee_percent`?

Comment (dpaiton) [reply elfpy/pricing_models/yieldspace.py:549]: Also, why do you need `abs`?

File: elfpy/pricing_models/yieldspace.py [hunk 26]
```diff
@@ -396,23 +385,12 @@ def calc_in_given_out(
...
-            # fee = (1 - p) * phi * d_y
-            fee = (1 - spot_price) * trade_fee_percent * d_bonds
-            logging.debug(
-                ("fee = (1 - spot_price) * _fee_percent * d_bonds = (1 - %g) * %g * %g = %g"),
-                spot_price,
-                trade_fee_percent,
-                d_bonds,
-                fee,
-            )
+            curve_fee = abs(d_bonds - without_fee_or_slippage) * trade_fee_percent
```
Comment (dpaiton) [inline elfpy/pricing_models/yieldspace.py:388]: hy do you need abs?

File: tests/solidity/test_fees.py [hunk 27]
```diff
@@ -0,0 +1,381 @@
...
+        )
+        _, wallet_deltas = self.hyperdrive.initialize(
+            wallet_address=self.alice.wallet.address,
+            contribution=self.contribution,
+            target_apr=self.target_apr,
+        )
+        self.alice.wallet.update(wallet_deltas)
+        self.market_state_before_open = self.hyperdrive.market_state.copy()
+
+    def assertAlmostEqual(self, first, second, places=None, msg=None, delta=None):
```
Comment (dpaiton) [inline tests/solidity/test_fees.py:76]: You should not have to cast anything to Decimal. The outputs of the calc functions (trade breakdown), as well as the market state should already be casting everything to a float.

File: tests/solidity/test_fees.py (hunk 15 above)
Comment (dpaiton) [reply tests/solidity/test_fees.py:95]: Ok, then can you rename this to `advance_time_and_set_share_price` or something like that?

File: elfpy/markets/hyperdrive/hyperdrive_market.py (hunk 24 above)
Comment (wakamex) [reply elfpy/markets/hyperdrive/hyperdrive_market.py]: done in 95eb80fe112de5f6a24b9cb3aebf1f720f95e83b

File: elfpy/pricing_models/yieldspace.py (hunk 25 above)
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py]: works correctly as suggested, implemented in 949ae8faae83acb3c18cc21996299a335c02de14

File: elfpy/pricing_models/yieldspace.py (hunk 26 above)
Comment (wakamex) [reply elfpy/pricing_models/yieldspace.py:385]: fixed in 36ec7f8093dd976989f9f9c157880796a8781a0d

File: tests/solidity/test_fees.py (hunk 27 above)
Comment (wakamex) [reply tests/solidity/test_fees.py]: everything passes without it lol
i swear this wasn't the case before
removed in 0968416ac0c76e40b2434dee977500c55265e22c

File: tests/solidity/test_fees.py (hunk 15 above)
Comment (wakamex) [reply tests/solidity/test_fees.py:84]: I renamed it to `advance_time` in a later commit. that matches solidity. they don't use share price explicitly anymore, so i'd rather leave that part out.

**Comment by dpaiton** [comment]
Looks good to me, thanks a lot for the hard work & multiple review iterations!
